lib/fixed-points.nix

```
@@ -394,4 +394,64 @@ rec {
     fix' (self: (rattrs self) // {
       ${extenderName} = f: makeExtensibleWithCustomName extenderName (extends f rattrs);
     });
+
+  /**
+    Convert to an extending function (overlay).
+
+    `toExtension` is the `toFunction` for extending functions (a.k.a. extensions or overlays).
+    It converts a non-function or a single-argument function to an extending function,
+    while returning a double-argument function as-is.
+
+    That is, it takes one of `x`, `prev: x`, or `final: prev: x`,
+    and returns `final: prev: x`, where `x` is not a function.
+
+    This function is extracted from the implementation of
+    the fixed-point arguments support of `stdenv.mkDerivation`.
+    It bridges the gap between `<pkg>.overrideAttrs`
+    before and after the overlay-style support,
+    as well as `config.packageOverrides` and `config.overlays` in `pkgs`.
+
+    # Inputs
+
+    `f0`
+    : The function or non-function to convert to an extending function.
+
+    # Type
+
+    ```
+    toExtension ::: (a | a -> a | a -> a -> a) -> a -> a -> a
+    a = AttrSet & !Function
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.fixedPoints.toExtension` usage example
+
+    ```nix
+    fix (final: { a = 0; c = final.a; })
+    => { a = 0; c = 0; };
+
+    fix (extends (toExtension { a = 1; b = 2; }) (final: { a = 0; c = final.a; }))
+    => { a = 1; b = 2; c = 1; };
+
+    fix (extends (toExtension (prev: { a = 1; b = prev.a; })) (final: { a = 0; c = final.a; }))
+    => { a = 1; b = 0; c = 1; };
+
+    fix (extends (toExtension (final: prev: { a = 1; b = prev.a; c = final.a + 1 })) (final: { a = 0; c = final.a; }))
+    => { a = 1; b = 0; c = 2; };
+    ```
+    :::
+  */
+  toExtension = f0:
+  if lib.isFunction f0 then
+    final: prev:
+      if lib.isFunction (f0 prev) then
+        # f0 is (final: prev: { ... })
+        f0 final prev
+      else
+        # f0 is (prev: { ... })
+        f0 prev
```

```suggestion
      let fPrev = f0 prev;
      if lib.isFunction (f0 prev) then
        # f0 is (final: prev: { ... })
        f0 final prev
      else
        # f0 is (prev: { ... })
        fPrev
```

`f0 prev` may (or may not) be expensive to recompute.
Let's be on the safe side.

I think you could rename `f0` to `f` as well, because there's no ambiguity with other `f`s. Or even `v`, because we don't really know whether `f` is even a function, as in the final case.